pkg/kubelet/userns_manager.go

```diff
@@ -0,0 +1,514 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kubelet
+
+import (
+	"encoding/json"
+	"fmt"
+	"math"
+	"os"
+	"sync"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
+	runtimeapi "k8s.io/cri-api/pkg/apis/runtime/v1"
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/pkg/features"
+	kubecontainer "k8s.io/kubernetes/pkg/kubelet/container"
+	utilstore "k8s.io/kubernetes/pkg/kubelet/util/store"
+	utilfs "k8s.io/kubernetes/pkg/util/filesystem"
+)
+
+// bitsDataElement is the number of bits in a bitArray.data element.
+const bitsDataElement = 32
+
+type bitArray struct {
+	data       []uint32
+	firstIndex int
+}
+
+func makeBitArray(size uint32) *bitArray {
+	m := bitArray{
+		data:       make([]uint32, (size+bitsDataElement-1)/bitsDataElement),
+		firstIndex: 0,
+	}
+	return &m
+}
+
+func (b *bitArray) set(index uint32) {
+	b.data[index/bitsDataElement] |= (uint32(1) << (index % bitsDataElement))
+}
+
+func (b *bitArray) isSet(index uint32) bool {
+	return (b.data[index/bitsDataElement]>>(index%bitsDataElement))&0x1 == 1
+}
+
+func (b *bitArray) findAndSetFirstZero() (uint32, error) {
+	for i := b.firstIndex; i < len(b.data); i++ {
+		// Check if all bits are used (all 1s).
+		if b.data[i] == math.MaxUint32 {
+			continue
+		}
+		for j := uint32(0); j < bitsDataElement; j++ {
+			if (b.data[i]>>j)&0x1 == 0 {
+				v := uint32(i)*bitsDataElement + j
+				b.set(v)
+				b.firstIndex = int(i)
+				return v, nil
+			}
+		}
+	}
+	return 0, fmt.Errorf("could not find an empty slot")
+}
+
+func (b *bitArray) clear(index uint32) {
+	i := index / bitsDataElement
+	if i < uint32(b.firstIndex) {
+		b.firstIndex = int(i)
+	}
+	b.data[i] &= ^(1 << (index % bitsDataElement))
+}
+
+// length for the user namespace to create (enough to fit 16 bits IDs).
+const userNsLength = (1 << 16)
```

Comment:
nit: Can we comment that this is 65536 :)

Reply:
Added

Is funny how different people want different things in the comment :)